crates/bevy_ecs/src/component.rs

```diff
@@ -378,6 +386,18 @@ pub trait Component: Send + Sync + 'static {
     /// A constant indicating the storage type used for this component.
     const STORAGE_TYPE: StorageType;
 
+    /// A marker type to assist Bevy with determining if this component is
+    /// mutable, or immutable. Mutable components will have [`ComponentMut`]
+    /// automatically implemented for them, while immutable components will
+    /// receive [`ComponentImmutable`].
+    ///
+    /// * For a component to be mutable, this type must be `Self`.
+    /// * For a component to be immutable, this type must be `()`.
+    ///
+    /// If you use any other type, this component will be immutable but will not
+    /// implement [`ComponentImmutable`].
+    type Mutable: 'static;
```

Comment:
IMO we should borrow from the approach taken by `StorageType`, and use a trait bound for marker types, rather than using `Self` / `()`. I think that's clearer, more consistent and it's much easier to extend.

Reply:
Gone with `Immutable` and `Mutable` as ZSTs implementing a new `ComponentMutability` trait. This also simplifies the boolean check for mutability by allowing the value to be stored on the ZST.